src/services/balancer/pools/joins/handlers/deep-pool-join.handler.ts

```diff
@@ -1,40 +1,105 @@
+import { balancer } from '@/lib/balancer.sdk';
+import { formatFixed, parseFixed } from '@ethersproject/bignumber';
+import { getAddress } from '@ethersproject/address';
+import { bnum } from '@/lib/utils';
 
 /**
  * Handles generalized joins for deep pools using SDK functions.
  */
+
+interface GeneralisedJoinResponse {
+  to: string;
+  callData: string;
+  minOut: string;
+  expectedOut: string;
+  // priceImpact: string;
+}
 export class DeepPoolJoinHandler implements JoinPoolHandler {
+  private lastGeneralisedJoinRes?: GeneralisedJoinResponse;
+
   constructor(
     public readonly pool: Ref<Pool>,
     public readonly sdk: BalancerSDK,
     public readonly gasPriceService: GasPriceService
   ) {}
 
-  async join({
+  async join(params: JoinParams): Promise<TransactionResponse> {
+    const { signer } = params;
+    await this.queryJoin(params);
+    if (!this.lastGeneralisedJoinRes) {
+      throw new Error('Could not query generalised join');
+    }
+    const { to, callData } = this.lastGeneralisedJoinRes;
+    return signer.sendTransaction({
+      to,
+      data: callData,
+    });
+  }
+
+  async queryJoin({
     amountsIn,
     tokensIn,
-    prices,
     signer,
     slippageBsp,
-  }: JoinParams): Promise<TransactionResponse> {
-    console.log([amountsIn, tokensIn, prices, signer, slippageBsp]);
-    throw new Error('To be implemented');
-  }
+    relayerSignature,
+  }: JoinParams): Promise<QueryOutput> {
+    const parsedAmountsIn: string[] = amountsIn.map(({ address, value }) => {
+      // Get the address in right casing style
+      const realAddress = getAddress(address);
+      const token = tokensIn[realAddress];
+      const parsedAmount = parseFixed(
+        value || '0',
+        token?.decimals ?? 18
+      ).toString();
+      return parsedAmount;
+    });
+
+    const tokenAddresses: string[] = amountsIn.map(({ address }) => address);
+    const signerAddress = await signer.getAddress();
+    const wrapLeafTokens = false;
+    const slippage = slippageBsp.toString();
+    const poolId = this.pool.value.id;
+
+    this.lastGeneralisedJoinRes = await balancer.pools
+      .generalisedJoin(
+        poolId,
+        tokenAddresses,
+        parsedAmountsIn,
+        signerAddress,
+        wrapLeafTokens,
+        slippage,
+        relayerSignature
+      )
+      .catch(err => {
+        console.error(err);
+        throw new Error(err);
+      });
+    console.log({ lastGeneralisedJoinRes: this.lastGeneralisedJoinRes });
+
+    if (!this.lastGeneralisedJoinRes) throw new Error('Not enough liquidity.');
+
+    const bptOut = formatFixed(
+      this.lastGeneralisedJoinRes.expectedOut,
+      this.pool.value.onchain?.decimals || 18
+    );
+    const priceImpact: number = bnum(
+      formatFixed(
+        // @ts-ignore-next-line -- priceImpact is part of the response, but type is missing
+        this.lastGeneralisedJoinRes.priceImpact,
+        this.pool.value.onchain?.decimals || 18
+      )
```

Comment:
If price impact is being returned in evm number format it will always be scaled with 18. You only use the decimals of the pool or token when you're dealing with amounts of the relevant token.

Reply:
Make sure this is the case with Bruno, but I'm pretty sure you can just use 18 here.